### PR TITLE
Add pointer indirection for write barrier calls

### PR DESCRIPTION
**Pull request by tenderlove**
This commit sets a function pointer on object space and switches the
write barrier function during incremental marking.  This means we can
eliminate a branch call at the price of indirection.

The write barrier is used *very* frequently, so it's important to make sure it's fast.  I tried this for a benchmark:

```ruby
class Thing
  def initialize a, b, c, d, e, f, g, h
    @a = a
    @b = b
    @c = c
    @d = d
    @e = e
    @f = f
    @g = g
    @h = h
  end
end

list = [Object.new, Object.new, Object.new, Object.new, Object.new, Object.new, Object.new, Object.new]

2000000.times do
  Thing.new(*list)
end
```

The instance variable writes will hit the write barrier, but I didn't see any difference in performance.  I'm *guessing* the branch predictor is right so frequently in the original version that I wouldn't see a difference between that an the jumping version.

Does anyone have any insight?  /cc @github/platform-systems

### CODE REVIEW

**Comment by tenderlove** [comment]
Here's another possible benchmark:

```ruby
require 'benchmark/ips'

GC.disable

class Thing
  attr_reader :omglol

  def go obj
    Benchmark.ips do |x|
      x.report("thing") do
        @omglol = obj
      end
    end
  end
end

t = Thing.new
t.go Object.new
p t.omglol
```

For which I still see no perf difference.

**Comment by vmg** [comment]
I think these results are reasonable. We're not really saving a branch here -- yes, the `if` is gone, but that function pointer is most definitely still branching. It just branches on the value of the `wb` variable (which is nothing but a memory address), as opposed to the boolean it was using before.

We still need to load the memory address from `wb` and then jump to it, which will either stall the pipeline, or be optimized by the CPU's predictor. I'm not even sure if the branch predictor will do this kind of arbitrary jumps (I assume it does or we'd see a perf regression here). Either way, the branch predictor is still working its ass off, and worse, the compiler can no longer inline the function calls we're doing to set the WB config.

So in summary: replacing an `if` with a function pointer will never give performance improvements. You would need to be replacing at the very least a `switch` to make up for the jump indirection, and even that seems unlikely, as GCC already compiles dense switches to jump tables.

Gotta go back to the drawing board and see what else can we do with write barriers. 😄

**Comment by tenderlove** [comment]
@vmg 💘💙💜💗💚❤💓💛💚💗 

I think we could save time if we moved the write barrier to the object and installed a different write barrier on old objects.  Only problem is that it would increase the size of the RVALUE, so I'm trying to think of some tricks.

**Comment by bkripto** [comment]
@VMG Get it V!
